File: src/gravai/recording/common/browser_persona.js
// Makes this browser look like Chrome on a Windows desktop, to whatever runs
// inside the page.
//
// The headers a persona rewrites are handled over CDP (user agent, client hints,
// timezone, locale); this file covers what only a script can see - the platform
// strings, the plugin list, the screen, the GPU, the installed voices and fonts.
// Values come from {{PERSONA}}, substituted by browser_persona.py.
//
// Two constraints this file has to respect, both learned the hard way:
//
//   * It is injected at document start into *every* frame, including the
//     opaque-origin iframes Meet creates while loading. Registering a blob URL
//     in one of those kills the renderer (see rtc_intercept.js), so nothing here
//     touches URL.createObjectURL, and every step is wrapped so a frame that
//     lacks an API skips it instead of throwing.
//   * A patch that is visible as a patch is worse than no patch. Anything
//     replaced here reports itself as native code when stringified.
(() => {
  "use strict";

  const P = {{PERSONA}};

  const safe = (fn) => {
    try {
      fn();
    } catch (_) {
      // A frame without this API, or one that froze it. Nothing to do.
    }
  };

  // --- keeping the patches from announcing themselves ---------------------
  //
  // Function.prototype.toString on a replaced getter would print its source,
  // which is a plainer tell than the value it hides. Every function installed
  // below is registered here and stringifies as native code instead.
  const nativeNames = new WeakMap();
  const nativeToString = Function.prototype.toString;

  const cloakedToString = new Proxy(nativeToString, {
    apply(target, thisArg, args) {
      const name = nativeNames.get(thisArg);
      if (name !== undefined) {
        return "function " + name + "() { [native code] }";
      }
      return Reflect.apply(target, thisArg, args);
    },
  });

  safe(() => {
    Function.prototype.toString = cloakedToString;
    nativeNames.set(cloakedToString, "toString");
    nativeNames.set(nativeToString, "toString");
  });

  const asNative = (fn, name) => {
    nativeNames.set(fn, name);
    return fn;
  };

  /** Replaces an accessor, leaving it configurable and native-looking. */
  const defineGetter = (target, prop, value) => {
    safe(() => {
      const getter = asNative(() => value, "get " + prop);
      Object.defineProperty(target, prop, {
        get: getter,
        set: undefined,
        configurable: true,
        enumerable: true,
      });
    });
  };

  /** Replaces a method, keeping the name Chrome would report for it. */
  const defineMethod = (target, prop, impl) => {
    safe(() => {
      const label = typeof prop === "string" ? prop : "[" + String(prop) + "]";
      Object.defineProperty(impl, "name", { value: label, configurable: true });
      asNative(impl, label);
      Object.defineProperty(target, prop, {
        value: impl,
        writable: true,
        configurable: true,
        enumerable: false,
      });
    });
  };

  // Properties live on the prototype in Chrome, not on the instance. Patching
  // the instance leaves the prototype's own descriptor visible one step up the
  // chain, which is exactly the sort of inconsistency that gets looked for.
  const navProto = Object.getPrototypeOf(navigator) || Navigator.prototype;
  const screenProto = Object.getPrototypeOf(screen) || Screen.prototype;

  // --- navigator ----------------------------------------------------------

  // --disable-blink-features=AutomationControlled already reports false here.
  // Only patch when it did not take, and to false rather than undefined: a real
  // Chrome has the property.
  if (navigator.webdriver !== false) {
    defineGetter(navProto, "webdriver", false);
  }

  defineGetter(navProto, "platform", P.platform);
  defineGetter(navProto, "vendor", P.vendor);
  defineGetter(navProto, "languages", Object.freeze(P.languages.slice()));
  defineGetter(navProto, "hardwareConcurrency", P.hardwareConcurrency);
  defineGetter(navProto, "deviceMemory", P.deviceMemory);
  defineGetter(navProto, "maxTouchPoints", P.maxTouchPoints);
  defineGetter(navProto, "pdfViewerEnabled", true);

  // A desktop Chrome reports a connection; a container that never had one is
  // conspicuous by the missing object.
  safe(() => {
    if (!navigator.connection && window.NetworkInformation) {
      const connection = Object.create(NetworkInformation.prototype);
      defineGetter(connection, "downlink", P.connection.downlink);
      defineGetter(connection, "effectiveType", P.connection.effectiveType);
      defineGetter(connection, "rtt", P.connection.rtt);
      defineGetter(connection, "saveData", false);
      defineGetter(navProto, "connection", connection);
    }
  });

  // --- plugins and mime types ---------------------------------------------
  //
  // Headless reports none. The five below are what Chrome ships on the desktop:
  // all five are the same built-in PDF viewer under different names, and they
  // are cross-linked with the two mime types, since a plugin list whose entries
  // do not resolve back through namedItem is its own tell.
  safe(() => {
    if (!window.Plugin || !window.PluginArray || navigator.plugins.length > 0) {
      return;
    }

    const mimeTypes = P.mimeTypes.map((spec) => {
      const mime = Object.create(MimeType.prototype);
      defineGetter(mime, "type", spec.type);
      defineGetter(mime, "suffixes", spec.suffixes);
      defineGetter(mime, "description", spec.description);
      return mime;
    });

    const plugins = P.plugins.map((spec) => {
      const plugin = Object.create(Plugin.prototype);
      defineGetter(plugin, "name", spec.name);
      defineGetter(plugin, "filename", spec.filename);
      defineGetter(plugin, "description", spec.description);
      defineGetter(plugin, "length", mimeTypes.length);
      mimeTypes.forEach((mime, index) => {
        defineGetter(plugin, String(index), mime);
        defineGetter(plugin, mime.type, mime);
      });
      defineMethod(plugin, "item", (index) => mimeTypes[index] || null);
      defineMethod(
        plugin,
        "namedItem",
        (name) => mimeTypes.find((mime) => mime.type === name) || null,
      );
      return plugin;
    });

    // Each mime type points back at the plugin that handles it.
    mimeTypes.forEach((mime) => defineGetter(mime, "enabledPlugin", plugins[0]));

    const collection = (proto, items, keyOf) => {
      const list = Object.create(proto);
      items.forEach((item, index) => {
        defineGetter(list, String(index), item);
        defineGetter(list, keyOf(item), item);
      });
      defineGetter(list, "length", items.length);
      defineMethod(list, "item", (index) => items[index] || null);
      defineMethod(
        list,
        "namedItem",
        (name) => items.find((item) => keyOf(item) === name) || null,
      );
      defineMethod(list, "refresh", () => undefined);
      defineMethod(list, Symbol.iterator, function* () {
        yield* items;
      });
      return list;
    };

    defineGetter(
      navProto,
      "plugins",
      collection(PluginArray.prototype, plugins, (plugin) => plugin.name),
    );
    defineGetter(
      navProto,
      "mimeTypes",
      collection(MimeTypeArray.prototype, mimeTypes, (mime) => mime.type),
    );
  });

  // --- screen and window geometry -----------------------------------------
  //
  // The browser is launched at the persona's size, so the inner dimensions are
  // already right; these are the desktop around it, which the container has none
  // of. availHeight is short by the height of the Windows taskbar.
  defineGetter(screenProto, "width", P.screen.width);
  defineGetter(screenProto, "height", P.screen.height);
  defineGetter(screenProto, "availWidth", P.screen.availWidth);
  defineGetter(screenProto, "availHeight", P.screen.availHeight);
  defineGetter(screenProto, "availLeft", 0);
  defineGetter(screenProto, "availTop", 0);
  defineGetter(screenProto, "colorDepth", P.screen.colorDepth);
  defineGetter(screenProto, "pixelDepth", P.screen.colorDepth);

  safe(() => {
    if (window.top === window.self) {
      // Only in the top frame: an iframe reports the values of the window that
      // holds it, and rewriting them there contradicts its own geometry.
      defineGetter(window, "outerWidth", P.screen.width);
      defineGetter(window, "outerHeight", P.screen.availHeight);
      defineGetter(window, "screenX", 0);
      defineGetter(window, "screenY", 0);
      defineGetter(window, "screenLeft", 0);
      defineGetter(window, "screenTop", 0);
    }
    defineGetter(window, "devicePixelRatio", P.devicePixelRatio);
  });

  // --- the GPU ------------------------------------------------------------
  //
  // The honest answer here is SwiftShader: this LXC has no /dev/dri, so every
  // pixel is drawn on the CPU. What follows reports the workstation's card
  // instead - the strings and the limits that go with a real D3D11 driver.
  safe(() => {
    // Omitted deliberately when the persona would rather answer honestly: a
    // claimed card whose pixels are drawn by SwiftShader is a claim that can be
    // checked by rendering, and a caught claim is worse than a plain answer.
    if (!P.webgl) return;
    const spoofed = {
      37445: P.webgl.unmaskedVendor, // UNMASKED_VENDOR_WEBGL
      37446: P.webgl.unmaskedRenderer, // UNMASKED_RENDERER_WEBGL
      3379: P.webgl.maxTextureSize, // MAX_TEXTURE_SIZE
      34024: P.webgl.maxTextureSize, // MAX_RENDERBUFFER_SIZE
      34076: P.webgl.maxTextureSize, // MAX_CUBE_MAP_TEXTURE_SIZE
    };

    for (const proto of [window.WebGLRenderingContext, window.WebGL2RenderingContext]) {
      if (!proto) continue;
      const original = proto.prototype.getParameter;
      defineMethod(proto.prototype, "getParameter", function getParameter(parameter) {
        if (parameter === 3386) {
          // MAX_VIEWPORT_DIMS, which is a typed array rather than a number.
          return new Int32Array(P.webgl.maxViewportDims);
        }
        if (parameter in spoofed) {
          return spoofed[parameter];
        }
        return original.call(this, parameter);
      });
    }
  });

  // WebGPU reports the adapter by name on a real desktop, and either nothing or
  // a software fallback here.
  safe(() => {
    if (!P.webgpu || !navigator.gpu || !window.GPUAdapter) return;
    const info = {
      vendor: P.webgpu.vendor,
      architecture: P.webgpu.architecture,
      device: "",
      description: P.webgpu.description,
    };
    if ("info" in GPUAdapter.prototype) {
      defineGetter(GPUAdapter.prototype, "info", info);
    }
    if (GPUAdapter.prototype.requestAdapterInfo) {
      defineMethod(GPUAdapter.prototype, "requestAdapterInfo", function requestAdapterInfo() {
        return Promise.resolve(info);
      });
    }
  });

  // --- window.chrome ------------------------------------------------------
  //
  // Real Chrome ships this; Chromium and headless leave holes in it. Only the
  // missing pieces are filled, so a genuine object is left alone.
  safe(() => {
    if (!window.chrome) {
      Object.defineProperty(window, "chrome", { value: {}, writable: true, configurable: true });
    }
    const chrome = window.chrome;
    if (!chrome.app) {
      chrome.app = {
        isInstalled: false,
        InstallState: { DISABLED: "disabled", INSTALLED: "installed", NOT_INSTALLED: "not_installed" },
        RunningState: { CANNOT_RUN: "cannot_run", READY_TO_RUN: "ready_to_run", RUNNING: "running" },
        getDetails: asNative(function getDetails() { return null; }, "getDetails"),
        getIsInstalled: asNative(function getIsInstalled() { return false; }, "getIsInstalled"),
        runningState: asNative(function runningState() { return "cannot_run"; }, "runningState"),
      };
    }
    if (!chrome.csi) {
      chrome.csi = asNative(function csi() {
        const timing = performance.timing || {};
        const start = timing.navigationStart || Date.now();
        return {
          startE: start,
          onloadT: timing.domContentLoadedEventEnd || start,
          pageT: performance.now(),
          tran: 15,
        };
      }, "csi");
    }
    if (!chrome.loadTimes) {
      chrome.loadTimes = asNative(function loadTimes() {
        const nav = performance.getEntriesByType("navigation")[0] || {};
        const start = (performance.timeOrigin || Date.now()) / 1000;
        return {
          requestTime: start,
          startLoadTime: start,
          commitLoadTime: start + (nav.responseStart || 0) / 1000,
          finishDocumentLoadTime: start + (nav.domContentLoadedEventEnd || 0) / 1000,
          finishLoadTime: start + (nav.loadEventEnd || 0) / 1000,
          firstPaintTime: start + (nav.responseEnd || 0) / 1000,
          firstPaintAfterLoadTime: 0,
          navigationType: "Other",
          wasFetchedViaSpdy: true,
          wasNpnNegotiated: true,
          npnNegotiatedProtocol: "h3",
          wasAlternateProtocolAvailable: false,
          connectionInfo: "h3",
        };
      }, "loadTimes");
    }
  });

  // --- permissions --------------------------------------------------------
  //
  // Headless answers 'prompt' for notifications while Notification.permission
  // says 'denied'. A desktop Chrome never disagrees with itself here.
  safe(() => {
    const permissions = navigator.permissions;
    if (!permissions || !permissions.query) return;
    const original = permissions.query;
    defineMethod(permissions, "query", function query(parameters) {
      if (parameters && parameters.name === "notifications") {
        return Promise.resolve({
          state: Notification.permission === "denied" ? "prompt" : Notification.permission,
          onchange: null,
        });
      }
      return original.call(permissions, parameters);
    });
  });

  safe(() => {
    if (window.Notification && Notification.permission === "denied") {
      defineGetter(Notification, "permission", "default");
    }
  });

  // --- installed voices ---------------------------------------------------
  //
  // A Windows desktop has the Microsoft voices; this container has no speech
  // engine at all, so getVoices returns an empty list - a cheap thing to check
  // and a definite one when it comes back empty.
  safe(() => {
    if (!window.speechSynthesis || speechSynthesis.getVoices().length > 0) return;
    const voices = P.voices.map((spec) => ({
      voiceURI: spec.name,
      name: spec.name,
      lang: spec.lang,
      localService: spec.localService,
      default: spec.default === true,
    }));
    defineMethod(speechSynthesis, "getVoices", function getVoices() {
      return voices.slice();
    });
  });

  // --- fonts --------------------------------------------------------------
  //
  // Only answers the question 'is this font installed'. Text measured on a
  // canvas is still drawn with whatever fonts this machine actually has, which
  // no script can change - see the note in browser_persona.py.
  safe(() => {
    if (!document.fonts || !document.fonts.check) return;
    const installed = new Set(P.fonts.map((font) => font.toLowerCase()));
    const original = document.fonts.check.bind(document.fonts);
    defineMethod(document.fonts, "check", function check(font, text) {
      const family = String(font)
        .replace(/^.*?\d+(\.\d+)?(px|pt|em|rem|%)\s+/, "")
        .replace(/['"]/g, "")
        .split(",")[0]
        .trim()
        .toLowerCase();
      if (installed.has(family)) {
        return true;
      }
      return original(font, text);
    });
  });
})();

File: src/gravai/recording/common/browser_persona.py
"""Presents the recorder's browser as Chrome on a Windows desktop.

Google refuses this container's browser at Meet's front door - the
`ResolveMeetingSpace` RPC answers 403 - while an ordinary Chrome on a Windows
machine on the same connection is offered the green room. This module is the
attempt to close the gap that a browser can close: every difference between the
two fingerprints that is a *string the page reports* rather than a fact about
the machine.

It works on three levels, because a persona that only changes one contradicts
itself on the others:

    headers      user agent and the full client-hint set, rewritten together
                 over CDP - Playwright's user_agent option leaves sec-ch-ua
                 alone, so a UA-only spoof says Windows and Linux at once
    scripts      browser_persona.js, injected at document start
    the process  a real Chrome binary rather than bundled Chromium, no
                 automation flag, and a window the size of the persona's screen

What it cannot change (even with JS):

    the renderer     this LXC has no /dev/dri, so drawing happens on the CPU in
                     SwiftShader. The reported strings are the workstation's
                     card; anything actually rasterised is not.
    font metrics     text measured on a canvas uses the 58 fonts installed here,
                     not Windows'. `document.fonts.check` is answered from the
                     persona, but `measureText` is answered by fontconfig.
    the TLS stack    Chrome's own, so this one is honest by construction.

Use it through `apply_persona`, which needs the page before it navigates:

    page = context.new_page()
    apply_persona(page, WINDOWS_CHROME)
    page.goto(meeting_url)
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field, replace
from functools import lru_cache
from importlib import resources
from typing import Any

# Injected into every frame; read from disk once per process.
_SCRIPT_PATH = resources.files("gravai.recording") / "common" / "browser_persona.js"


@dataclass(frozen=True)
class BrowserPersona:
    """One coherent identity: what the browser says over the wire and in-page."""

    name: str
    user_agent: str
    ua_metadata: dict[str, Any]
    accept_language: str
    locale: str
    timezone: str
    # navigator.platform, which client hints do not cover.
    js_platform: str
    viewport: dict[str, int]
    launch_args: list[str]
    # Playwright adds these; a real Chrome has none of them.
    ignore_default_args: list[str]
    # Everything browser_persona.js reads, serialised into it as {{PERSONA}}.
    js_config: dict[str, Any] = field(default_factory=dict)


_PDF_PLUGINS = [
    # Chrome ships one PDF viewer under five names. The list is fixed: a count
    # that is not five, from a browser claiming to be Chrome, is itself a tell.
    {"name": "PDF Viewer", "filename": "internal-pdf-viewer", "description": "Portable Document Format"},
    {"name": "Chrome PDF Viewer", "filename": "internal-pdf-viewer", "description": "Portable Document Format"},
    {"name": "Chromium PDF Viewer", "filename": "internal-pdf-viewer", "description": "Portable Document Format"},
    {"name": "Microsoft Edge PDF Viewer", "filename": "internal-pdf-viewer", "description": "Portable Document Format"},
    {"name": "WebKit built-in PDF", "filename": "internal-pdf-viewer", "description": "Portable Document Format"},
]

_PDF_MIME_TYPES = [
    {"type": "application/pdf", "suffixes": "pdf", "description": "Portable Document Format"},
    {"type": "text/pdf", "suffixes": "pdf", "description": "Portable Document Format"},
]

# The speech engines a Windows install has and a Debian container does not.
_WINDOWS_VOICES = [
    {"name": "Microsoft David - English (United States)", "lang": "en-US", "localService": True, "default": True},
    {"name": "Microsoft Mark - English (United States)", "lang": "en-US", "localService": True},
    {"name": "Microsoft Zira - English (United States)", "lang": "en-US", "localService": True},
    {"name": "Microsoft Maria - Portuguese (Brazil)", "lang": "pt-BR", "localService": True},
    {"name": "Microsoft Daniel - Portuguese (Brazil)", "lang": "pt-BR", "localService": True},
    {"name": "Google US English", "lang": "en-US", "localService": False},
    {"name": "Google português do Brasil", "lang": "pt-BR", "localService": False},
]

# Answers 'is this installed' for the fonts that ship with Windows and are
# absent here. Enumeration by name is what this covers; measured width is not.
_WINDOWS_FONTS = [
    "Segoe UI", "Segoe UI Emoji", "Segoe UI Symbol", "Segoe UI Historic",
    "Arial", "Arial Black", "Bahnschrift", "Calibri", "Cambria", "Cambria Math",
    "Candara", "Comic Sans MS", "Consolas", "Constantia", "Corbel",
    "Courier New", "Ebrima", "Franklin Gothic Medium", "Gabriola", "Gadugi",
    "Georgia", "HoloLens MDL2 Assets", "Impact", "Ink Free",
    "Javanese Text", "Leelawadee UI", "Lucida Console", "Lucida Sans Unicode",
    "Malgun Gothic", "Marlett", "Microsoft Himalaya", "Microsoft JhengHei",
    "Microsoft New Tai Lue", "Microsoft PhagsPa", "Microsoft Sans Serif",
    "Microsoft Tai Le", "Microsoft YaHei", "Microsoft Yi Baiti",
    "MingLiU-ExtB", "Mongolian Baiti", "MS Gothic", "MV Boli",
    "Myanmar Text", "Nirmala UI", "Palatino Linotype", "Segoe MDL2 Assets",
    "Segoe Print", "Segoe Script", "SimSun", "Sitka", "Sylfaen", "Symbol",
    "Tahoma", "Times New Roman", "Trebuchet MS", "Verdana", "Webdings",
    "Wingdings", "Yu Gothic",
]

# Chrome 151 as it ships on Windows 11. platformVersion is the client-hint
# encoding of the OS: Windows 11 reports 15.0.0, not 10.
_CHROME_VERSION = "151.0.7922.137"
_CHROME_MAJOR = "151"

WINDOWS_CHROME = BrowserPersona(
    name="windows-chrome",
    user_agent=(
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        f"(KHTML, like Gecko) Chrome/{_CHROME_MAJOR}.0.0.0 Safari/537.36"
    ),
    ua_metadata={
        # The GREASE brand is replaced at apply time with whatever this Chrome
        # really sends; hard-coding it guesses at a value that rotates by
        # version, and a wrong guess is worse than the Linux string it replaces.
        "brands": [
            {"brand": "Not;A=Brand", "version": "99"},
            {"brand": "Google Chrome", "version": _CHROME_MAJOR},
            {"brand": "Chromium", "version": _CHROME_MAJOR},
        ],
        "fullVersionList": [
            {"brand": "Not;A=Brand", "version": "99.0.0.0"},
            {"brand": "Google Chrome", "version": _CHROME_VERSION},
            {"brand": "Chromium", "version": _CHROME_VERSION},
        ],
        "fullVersion": _CHROME_VERSION,
        "platform": "Windows",
        "platformVersion": "15.0.0",
        "architecture": "x86",
        "bitness": "64",
        "model": "",
        "mobile": False,
        "wow64": False,
    },
    accept_language="en-US,en;q=0.9",
    locale="en-US",
    # The workstation's zone. A browser claiming Windows from a machine sitting
    # in UTC is a mismatch anyone can check with one call to Intl.
    timezone="America/Sao_Paulo",
    js_platform="Win32",
    viewport={"width": 1920, "height": 1080},
    launch_args=[
        "--disable-blink-features=AutomationControlled",
        "--window-size=1920,1080",
        "--window-position=0,0",
        "--lang=en-US",
        "--force-device-scale-factor=1",
        "--force-color-profile=srgb",
    ],
    ignore_default_args=[
        "--enable-automation",
        "--disable-background-networking",
        "--disable-component-update",
        "--disable-default-apps",
        "--disable-extensions",
        "--disable-sync",
        "--mute-audio",
    ],
    js_config={
        "platform": "Win32",
        "vendor": "Google Inc.",
        "languages": ["en-US", "en"],
        "hardwareConcurrency": 8,
        "deviceMemory": 8,
        "maxTouchPoints": 0,
        "devicePixelRatio": 1,
        "connection": {"downlink": 10, "effectiveType": "4g", "rtt": 50},
        "plugins": _PDF_PLUGINS,
        "mimeTypes": _PDF_MIME_TYPES,
        "screen": {
            "width": 1920,
            "height": 1080,
            "availWidth": 1920,
            # 1080 less the Windows 11 taskbar.
            "availHeight": 1032,
            "colorDepth": 24,
        },
        "webgl": {
            "unmaskedVendor": "Google Inc. (AMD)",
            "unmaskedRenderer": (
                "ANGLE (AMD, AMD Radeon RX 9070 XT (0x00007550) Direct3D11 vs_5_0 ps_5_0, "
                "D3D11-32.0.21023.1017)"
            ),
            # SwiftShader's limits differ from a D3D11 driver's, and they are one
            # getParameter call away from the renderer string that was spoofed.
            "maxTextureSize": 16384,
            "maxViewportDims": [32767, 32767],
        },
        "webgpu": {
            "vendor": "amd",
            "architecture": "rdna-4",
            "description": "AMD Radeon RX 9070 XT",
        },
        "voices": _WINDOWS_VOICES,
        "fonts": _WINDOWS_FONTS,
    },
)

# The same persona, minus the graphics card it does not have. Worth having as a
# named variant rather than a flag: a claimed renderer can be checked against
# what actually gets drawn, so which of the two is more convincing is an open
# question that gets re-tested, not a setting anyone should have to guess at.
WINDOWS_CHROME_HONEST_GPU = replace(
    WINDOWS_CHROME,
    name="windows-chrome-honest-gpu",
    js_config={
        key: value
        for key, value in WINDOWS_CHROME.js_config.items()
        if key not in ("webgl", "webgpu")
    },
)

PERSONAS = {
    persona.name: persona for persona in (WINDOWS_CHROME, WINDOWS_CHROME_HONEST_GPU)
}


@lru_cache(maxsize=4)
def _script_template() -> str:
    return _SCRIPT_PATH.read_text(encoding="utf-8")


def init_script(persona: BrowserPersona) -> str:
    """The document-start script for this persona."""
    return _script_template().replace("{{PERSONA}}", json.dumps(persona.js_config))


def launch_kwargs(persona: BrowserPersona, headless: bool = True, channel: str | None = "chrome") -> dict:
    """Launch options for `chromium.launch`.

    `channel="chrome"` runs the real Google Chrome rather than the bundled
    Chromium: a persona claiming to be Chrome from a Chromium build disagrees
    with itself in more places than can be patched by hand.
    """
    kwargs: dict[str, Any] = {
        "headless": headless,
        "args": list(persona.launch_args),
        "ignore_default_args": list(persona.ignore_default_args),
    }
    if channel:
        kwargs["channel"] = channel
    return kwargs


def context_kwargs(persona: BrowserPersona) -> dict:
    """Context options that belong to the persona rather than to the caller."""
    return {
        "locale": persona.locale,
        "timezone_id": persona.timezone,
        "viewport": dict(persona.viewport),
        "user_agent": persona.user_agent,
    }


def apply_persona(page, persona: BrowserPersona = WINDOWS_CHROME) -> None:
    """Applies the header-level half of the persona to one page.

    Must run before the page navigates: the overrides are read when a request is
    built, and a document already fetched keeps the headers it was fetched with.

    The script half is added to the *context* by the caller, so that it also
    covers frames and pages opened later:

        context.add_init_script(init_script(persona))
    """
    session = page.context.new_cdp_session(page)

    metadata = json.loads(json.dumps(persona.ua_metadata))
    _use_real_brands(page, metadata)

    session.send(
        "Emulation.setUserAgentOverride",
        {
            "userAgent": persona.user_agent,
            "acceptLanguage": persona.accept_language,
            "platform": persona.js_platform,
            "userAgentMetadata": metadata,
        },
    )
    # Set here as well as on the context: a context created without them, or one
    # this function is handed after the fact, still ends up consistent.
    for command, params in (
        ("Emulation.setTimezoneOverride", {"timezoneId": persona.timezone}),
        ("Emulation.setLocaleOverride", {"locale": persona.locale}),
    ):
        try:
            session.send(command, params)
        except Exception:
            # Already overridden to the same value by the context; harmless.
            pass


def _use_real_brands(page, metadata: dict) -> None:
    """Keeps this Chrome's own brand list, relabelled for the persona.

    Chrome pads `sec-ch-ua` with a deliberately varying junk brand ('Not;A=Brand'
    and friends), whose spelling and position change between versions. Reading
    the browser's real list and swapping only the version numbers is exact,
    where writing the list out by hand is a guess that dates.
    """
    try:
        brands = page.evaluate(
            "() => navigator.userAgentData && navigator.userAgentData.brands"
        )
    except Exception:
        return
    if not brands:
        return

    full_versions = {entry["brand"]: entry["version"] for entry in metadata["fullVersionList"]}
    metadata["brands"] = [dict(entry) for entry in brands]
    metadata["fullVersionList"] = [
        {
            "brand": entry["brand"],
            # A brand this persona has no full version for is the junk one,
            # whose full form is its major with three zeroes.
            "version": full_versions.get(entry["brand"], f"{entry['version']}.0.0.0"),
        }
        for entry in brands
    ]
